### DISCUSSION TITLE
List of all locales and corresponding languages?

### DISCUSSION THREAD

**Discussion by web-programmer-here** [js-fns/date-fns#2724] [Q&A]
Is there any list of locales and the corresponding languages?
for example

```shell
af        Afrikaans
ar-DZ     Algerian Arabic
and so on
```

**Comment by fturmel** [comment]
Nothing official that I know of, but you're in luck. I already had a list at hand that was generated using the jdsocs fields of each locale.

It could be a good opportunity for the community to review these fields and bring any corrections or improvements. I am aware of at least three issues:
- `fr-CH` should be `French locale (Switzerland)`, fix is part of #2626
- `nl-BE` should be something like ~`Dutch/Fleming (Belgium)`~ `Dutch (Belgium)`
- `ar-SA` should be `Saudi Arabic`, there is a typo in there

This is the data as of version 2.25.0:

<details><summary>Markdown table</summary>

| code    | ISO 639-2 code | summary                         | language            |
|---------|-----------|---------------------------------|---------------------|
| af      | afr       | Afrikaans locale                | Afrikaans           |
| ar-DZ   | ara       | Arabic locale (Algerian Arabic) | Algerian Arabic     |
| ar-MA   | ara       | Arabic locale (Moroccan Arabic) | Moroccan Arabic     |
| ar-SA   | ara       | Arabic locale (Sauid Arabic)    | Arabic              |
| ar-TN   | ara       | Arabic locale (Tunisian Arabic) | Arabic              |
| az      | aze       | Azerbaijani locale              | Azerbaijani         |
| be      | bel       | Belarusian locale               | Belarusian          |
| bg      | bul       | Bulgarian locale                | Bulgarian           |
| bn      | ben       | Bengali locale                  | Bengali             |
| bs      | bos       | Bosnian locale                  | Bosnian             |
| ca      | cat       | Catalan locale                  | Catalan             |
| cs      | ces       | Czech locale                    | Czech               |
| cy      | cym       | Welsh locale                    | Welsh               |
| da      | dan       | Danish locale                   | Danish              |
| de      | deu       | German locale                   | German              |
| de-AT   | deu       | German locale (Austria)         | German              |
| el      | ell       | Greek locale                    | Greek               |
| en-AU   | eng       | English locale (Australia)      | English             |
| en-CA   | eng       | English locale (Canada)         | English             |
| en-GB   | eng       | English locale (United Kingdom) | English             |
| en-IN   | eng       | English locale (India)          | English             |
| en-NZ   | eng       | English locale (New Zealand)    | English             |
| en-US   | eng       | English locale (United States)  | English             |
| en-ZA   | eng       | English locale (South Africa)   | English             |
| eo      | epo       | Esperanto locale                | Esperanto           |
| es      | spa       | Spanish locale                  | Spanish             |
| et      | est       | Estonian locale                 | Estonian            |
| eu      | eus       | Basque locale                   | Basque              |
| fa-IR   | ira       | Persian/Farsi locale (Iran)     | Persian             |
| fi      | fin       | Finnish locale                  | Finnish             |
| fr      | fra       | French locale                   | French              |
| fr-CA   | fra       | French locale (Canada)          | French              |
| fr-CH   | fra       | French locale                   | French              |
| gd      | gla       | Scottish Gaelic                 | Scottish Gaelic     |
| gl      | glg       | Galician locale                 | Galician            |
| gu      | guj       | Gujarati locale (India)         | Gujarati            |
| he      | heb       | Hebrew locale                   | Hebrew              |
| hi      | hin       | Hindi locale (India)            | Hindi               |
| hr      | hrv       | Croatian locale                 | Croatian            |
| ht      | hat       | Haitian Creole locale           | Haitian Creole      |
| hu      | hun       | Hungarian locale                | Hungarian           |
| hy      | arm       | Armenian locale                 | Armenian            |
| id      | ind       | Indonesian locale               | Indonesian          |
| is      | isl       | Icelandic locale                | Icelandic           |
| it      | ita       | Italian locale                  | Italian             |
| ja      | jpn       | Japanese locale                 | Japanese            |
| ja-Hira | jpn       | Japanese (Hiragana) locale      | Japanese (Hiragana) |
| ka      | geo       | Georgian locale                 | Georgian            |
| kk      | kaz       | Kazakh locale                   | Kazakh              |
| kn      | kan       | Kannada locale (India)          | Kannada             |
| ko      | kor       | Korean locale                   | Korean              |
| lb      | ltz       | Luxembourgish locale            | Luxembourgish       |
| lt      | lit       | Lithuanian locale               | Lithuanian          |
| lv      | lav       | Latvian locale (Latvia)         | Latvian             |
| mk      | mkd       | Macedonian locale               | Macedonian          |
| mn      | mon       | Mongolian locale                | Mongolian           |
| ms      | msa       | Malay locale                    | Malay               |
| mt      | mlt       | Maltese locale                  | Maltese             |
| nb      | nob       | Norwegian Bokmål locale         | Norwegian Bokmål    |
| nl      | nld       | Dutch locale                    | Dutch               |
| nl-BE   | nld       | Dutch locale                    | Dutch               |
| nn      | nno       | Norwegian Nynorsk locale        | Norwegian Nynorsk   |
| pl      | pol       | Polish locale                   | Polish              |
| pt      | por       | Portuguese locale               | Portuguese          |
| pt-BR   | por       | Portuguese locale (Brazil)      | Portuguese          |
| ro      | ron       | Romanian locale                 | Romanian            |
| ru      | rus       | Russian locale                  | Russian             |
| sk      | slk       | Slovak locale                   | Slovak              |
| sl      | slv       | Slovenian locale                | Slovenian           |
| sq      | sqi       | Albanian locale                 | Shqip               |
| sr      | srp       | Serbian cyrillic locale         | Serbian             |
| sr-Latn | srp       | Serbian latin locale            | Serbian             |
| sv      | swe       | Swedish locale                  | Swedish             |
| ta      | tam       | Tamil locale (India)            | Tamil               |
| te      | tel       | Telugu locale                   | Telugu              |
| th      | tha       | Thai locale                     | Thai                |
| tr      | tur       | Turkish locale                  | Turkish             |
| ug      | uig       | Uighur locale                   | Uighur              |
| uk      | ukr       | Ukrainian locale                | Ukrainian           |
| uz      | uzb       | Uzbek locale                    | Uzbek               |
| vi      | vie       | Vietnamese locale (Vietnam)     | Vietnamese          |
| zh-CN   | zho       | Chinese Simplified locale       | Chinese Simplified  |
| zh-TW   | zho       | Chinese Traditional locale      | Chinese Traditional |
</details>

<details><summary>JSON</summary>

```json
[
  {
    "code": "af",
    "iso_639_2": "afr",
    "summary": "Afrikaans locale",
    "language": "Afrikaans"
  },
  {
    "code": "ar-DZ",
    "iso_639_2": "ara",
    "summary": "Arabic locale (Algerian Arabic)",
    "language": "Algerian Arabic"
  },
  {
    "code": "ar-MA",
    "iso_639_2": "ara",
    "summary": "Arabic locale (Moroccan Arabic)",
    "language": "Moroccan Arabic"
  },
  {
    "code": "ar-SA",
    "iso_639_2": "ara",
    "summary": "Arabic locale (Sauid Arabic)",
    "language": "Arabic"
  },
  {
    "code": "ar-TN",
    "iso_639_2": "ara",
    "summary": "Arabic locale (Tunisian Arabic)",
    "language": "Arabic"
  },
  {
    "code": "az",
    "iso_639_2": "aze",
    "summary": "Azerbaijani locale",
    "language": "Azerbaijani"
  },
  {
    "code": "be",
    "iso_639_2": "bel",
    "summary": "Belarusian locale",
    "language": "Belarusian"
  },
  {
    "code": "bg",
    "iso_639_2": "bul",
    "summary": "Bulgarian locale",
    "language": "Bulgarian"
  },
  {
    "code": "bn",
    "iso_639_2": "ben",
    "summary": "Bengali locale",
    "language": "Bengali"
  },
  {
    "code": "bs",
    "iso_639_2": "bos",
    "summary": "Bosnian locale",
    "language": "Bosnian"
  },
  {
    "code": "ca",
    "iso_639_2": "cat",
    "summary": "Catalan locale",
    "language": "Catalan"
  },
  {
    "code": "cs",
    "iso_639_2": "ces",
    "summary": "Czech locale",
    "language": "Czech"
  },
  {
    "code": "cy",
    "iso_639_2": "cym",
    "summary": "Welsh locale",
    "language": "Welsh"
  },
  {
    "code": "da",
    "iso_639_2": "dan",
    "summary": "Danish locale",
    "language": "Danish"
  },
  {
    "code": "de",
    "iso_639_2": "deu",
    "summary": "German locale",
    "language": "German"
  },
  {
    "code": "de-AT",
    "iso_639_2": "deu",
    "summary": "German locale (Austria)",
    "language": "German"
  },
  {
    "code": "el",
    "iso_639_2": "ell",
    "summary": "Greek locale",
    "language": "Greek"
  },
  {
    "code": "en-AU",
    "iso_639_2": "eng",
    "summary": "English locale (Australia)",
    "language": "English"
  },
  {
    "code": "en-CA",
    "iso_639_2": "eng",
    "summary": "English locale (Canada)",
    "language": "English"
  },
  {
    "code": "en-GB",
    "iso_639_2": "eng",
    "summary": "English locale (United Kingdom)",
    "language": "English"
  },
  {
    "code": "en-IN",
    "iso_639_2": "eng",
    "summary": "English locale (India)",
    "language": "English"
  },
  {
    "code": "en-NZ",
    "iso_639_2": "eng",
    "summary": "English locale (New Zealand)",
    "language": "English"
  },
  {
    "code": "en-US",
    "iso_639_2": "eng",
    "summary": "English locale (United States)",
    "language": "English"
  },
  {
    "code": "en-ZA",
    "iso_639_2": "eng",
    "summary": "English locale (South Africa)",
    "language": "English"
  },
  {
    "code": "eo",
    "iso_639_2": "epo",
    "summary": "Esperanto locale",
    "language": "Esperanto"
  },
  {
    "code": "es",
    "iso_639_2": "spa",
    "summary": "Spanish locale",
    "language": "Spanish"
  },
  {
    "code": "et",
    "iso_639_2": "est",
    "summary": "Estonian locale",
    "language": "Estonian"
  },
  {
    "code": "eu",
    "iso_639_2": "eus",
    "summary": "Basque locale",
    "language": "Basque"
  },
  {
    "code": "fa-IR",
    "iso_639_2": "ira",
    "summary": "Persian/Farsi locale (Iran)",
    "language": "Persian"
  },
  {
    "code": "fi",
    "iso_639_2": "fin",
    "summary": "Finnish locale",
    "language": "Finnish"
  },
  {
    "code": "fr",
    "iso_639_2": "fra",
    "summary": "French locale",
    "language": "French"
  },
  {
    "code": "fr-CA",
    "iso_639_2": "fra",
    "summary": "French locale (Canada)",
    "language": "French"
  },
  {
    "code": "fr-CH",
    "iso_639_2": "fra",
    "summary": "French locale",
    "language": "French"
  },
  {
    "code": "gd",
    "iso_639_2": "gla",
    "summary": "Scottish Gaelic",
    "language": "Scottish Gaelic"
  },
  {
    "code": "gl",
    "iso_639_2": "glg",
    "summary": "Galician locale",
    "language": "Galician"
  },
  {
    "code": "gu",
    "iso_639_2": "guj",
    "summary": "Gujarati locale (India)",
    "language": "Gujarati"
  },
  {
    "code": "he",
    "iso_639_2": "heb",
    "summary": "Hebrew locale",
    "language": "Hebrew"
  },
  {
    "code": "hi",
    "iso_639_2": "hin",
    "summary": "Hindi locale (India)",
    "language": "Hindi"
  },
  {
    "code": "hr",
    "iso_639_2": "hrv",
    "summary": "Croatian locale",
    "language": "Croatian"
  },
  {
    "code": "ht",
    "iso_639_2": "hat",
    "summary": "Haitian Creole locale",
    "language": "Haitian Creole"
  },
  {
    "code": "hu",
    "iso_639_2": "hun",
    "summary": "Hungarian locale",
    "language": "Hungarian"
  },
  {
    "code": "hy",
    "iso_639_2": "arm",
    "summary": "Armenian locale",
    "language": "Armenian"
  },
  {
    "code": "id",
    "iso_639_2": "ind",
    "summary": "Indonesian locale",
    "language": "Indonesian"
  },
  {
    "code": "is",
    "iso_639_2": "isl",
    "summary": "Icelandic locale",
    "language": "Icelandic"
  },
  {
    "code": "it",
    "iso_639_2": "ita",
    "summary": "Italian locale",
    "language": "Italian"
  },
  {
    "code": "ja",
    "iso_639_2": "jpn",
    "summary": "Japanese locale",
    "language": "Japanese"
  },
  {
    "code": "ja-Hira",
    "iso_639_2": "jpn",
    "summary": "Japanese (Hiragana) locale",
    "language": "Japanese (Hiragana)"
  },
  {
    "code": "ka",
    "iso_639_2": "geo",
    "summary": "Georgian locale",
    "language": "Georgian"
  },
  {
    "code": "kk",
    "iso_639_2": "kaz",
    "summary": "Kazakh locale",
    "language": "Kazakh"
  },
  {
    "code": "kn",
    "iso_639_2": "kan",
    "summary": "Kannada locale (India)",
    "language": "Kannada"
  },
  {
    "code": "ko",
    "iso_639_2": "kor",
    "summary": "Korean locale",
    "language": "Korean"
  },
  {
    "code": "lb",
    "iso_639_2": "ltz",
    "summary": "Luxembourgish locale",
    "language": "Luxembourgish"
  },
  {
    "code": "lt",
    "iso_639_2": "lit",
    "summary": "Lithuanian locale",
    "language": "Lithuanian"
  },
  {
    "code": "lv",
    "iso_639_2": "lav",
    "summary": "Latvian locale (Latvia)",
    "language": "Latvian"
  },
  {
    "code": "mk",
    "iso_639_2": "mkd",
    "summary": "Macedonian locale",
    "language": "Macedonian"
  },
  {
    "code": "mn",
    "iso_639_2": "mon",
    "summary": "Mongolian locale",
    "language": "Mongolian"
  },
  {
    "code": "ms",
    "iso_639_2": "msa",
    "summary": "Malay locale",
    "language": "Malay"
  },
  {
    "code": "mt",
    "iso_639_2": "mlt",
    "summary": "Maltese locale",
    "language": "Maltese"
  },
  {
    "code": "nb",
    "iso_639_2": "nob",
    "summary": "Norwegian Bokmål locale",
    "language": "Norwegian Bokmål"
  },
  {
    "code": "nl",
    "iso_639_2": "nld",
    "summary": "Dutch locale",
    "language": "Dutch"
  },
  {
    "code": "nl-BE",
    "iso_639_2": "nld",
    "summary": "Dutch locale",
    "language": "Dutch"
  },
  {
    "code": "nn",
    "iso_639_2": "nno",
    "summary": "Norwegian Nynorsk locale",
    "language": "Norwegian Nynorsk"
  },
  {
    "code": "pl",
    "iso_639_2": "pol",
    "summary": "Polish locale",
    "language": "Polish"
  },
  {
    "code": "pt",
    "iso_639_2": "por",
    "summary": "Portuguese locale",
    "language": "Portuguese"
  },
  {
    "code": "pt-BR",
    "iso_639_2": "por",
    "summary": "Portuguese locale (Brazil)",
    "language": "Portuguese"
  },
  {
    "code": "ro",
    "iso_639_2": "ron",
    "summary": "Romanian locale",
    "language": "Romanian"
  },
  {
    "code": "ru",
    "iso_639_2": "rus",
    "summary": "Russian locale",
    "language": "Russian"
  },
  {
    "code": "sk",
    "iso_639_2": "slk",
    "summary": "Slovak locale",
    "language": "Slovak"
  },
  {
    "code": "sl",
    "iso_639_2": "slv",
    "summary": "Slovenian locale",
    "language": "Slovenian"
  },
  {
    "code": "sq",
    "iso_639_2": "sqi",
    "summary": "Albanian locale",
    "language": "Shqip"
  },
  {
    "code": "sr",
    "iso_639_2": "srp",
    "summary": "Serbian cyrillic locale",
    "language": "Serbian"
  },
  {
    "code": "sr-Latn",
    "iso_639_2": "srp",
    "summary": "Serbian latin locale",
    "language": "Serbian"
  },
  {
    "code": "sv",
    "iso_639_2": "swe",
    "summary": "Swedish locale",
    "language": "Swedish"
  },
  {
    "code": "ta",
    "iso_639_2": "tam",
    "summary": "Tamil locale (India)",
    "language": "Tamil"
  },
  {
    "code": "te",
    "iso_639_2": "tel",
    "summary": "Telugu locale",
    "language": "Telugu"
  },
  {
    "code": "th",
    "iso_639_2": "tha",
    "summary": "Thai locale",
    "language": "Thai"
  },
  {
    "code": "tr",
    "iso_639_2": "tur",
    "summary": "Turkish locale",
    "language": "Turkish"
  },
  {
    "code": "ug",
    "iso_639_2": "uig",
    "summary": "Uighur locale",
    "language": "Uighur"
  },
  {
    "code": "uk",
    "iso_639_2": "ukr",
    "summary": "Ukrainian locale",
    "language": "Ukrainian"
  },
  {
    "code": "uz",
    "iso_639_2": "uzb",
    "summary": "Uzbek locale",
    "language": "Uzbek"
  },
  {
    "code": "vi",
    "iso_639_2": "vie",
    "summary": "Vietnamese locale (Vietnam)",
    "language": "Vietnamese"
  },
  {
    "code": "zh-CN",
    "iso_639_2": "zho",
    "summary": "Chinese Simplified locale",
    "language": "Chinese Simplified"
  },
  {
    "code": "zh-TW",
    "iso_639_2": "zho",
    "summary": "Chinese Traditional locale",
    "language": "Chinese Traditional"
  }
]

```
</details>

**Reply by bartduisters** [reply]
Dutch (Belgium) is what it should refer to in my opinion. Flemish (not Fleming) is not a language nor a dialect.

**Reply by fturmel** [reply]
Thank you, I will fix it in the PR below. Any thoughts on CLDR listing nl_BE as Flemish in the English column?

https://unicode-org.github.io/cldr-staging/charts/40/summary/nl.html

<img width="763" alt="Screen Shot 2021-12-11 at 6 01 50 PM" src="https://user-images.githubusercontent.com/4009209/145694195-dcd9360a-c784-45e1-907a-3f4ea1d854b8.png">

**Reply by bartduisters** [reply]
Yes, my thoughts are that CLDR has it wrong as well.

I've put a comment on the commit that introduced the change: https://github.com/date-fns/date-fns/commit/20963fe8614b5c8017d2dc7f6ba95ccf48e922b7

**Comment by fturmel** [comment]
For anyone interested in this topic, keep track of PR #2838. If merged, it will dynamically generate a page in the documentation that lists all available locales for each release.

**Comment by tan75** [comment]
Hi all, 

brilliant idea!
We will discuss it this coming week in our release meeting.

**Comment by mamiabudat** [comment]
I went to the repo, and found the folders named according to the languages supported. Documentation could be better
https://github.com/date-fns/date-fns/tree/main/src/locale

**Comment by suranjanz** [comment]
in  .net maui 2022 visualstudio its possible to get the locales using the TextToSpeech  even if you dont plan to use it beyond this ability to
get all country names and codes..... then dispose it

Locale MyLocale;
string TextFeedBack="Country Names ";
 var x = await TextToSpeech.GetLocalesAsync().ConfigureAwait(true);  //otherwise it will continue before result and foreach fails
 
 foreach (var locale in x)
 {
     LanguagePicker.Items.Add(locale.Name);
     TextFeedBack += locale.Country + " ";
 }
 if (LanguagePicker.Items.Count > 0)
 {
     LanguagePicker.SelectedIndex = LanguagePicker.Items.Count;  //last one is "EN UK"
    .MyLocale = LanguagePicker.Items.Single (x => x.Name == LanguagePicker.SelectedItem.ToString());
 }
